Properly colour links in FullscreenOverlays

thewildtree opened the pull request:
Fixes #7822.

Small change code-wise but brings a big improvement in the overall look and feel of `FullscreenOverlay`s. Only changes the idle colour, links still turn yellow when hovering.

## Comparisons

### Beatmap score display:

before:
![image](https://user-images.githubusercontent.com/27722894/75554861-70303800-5a3b-11ea-9e00-59325d10ac93.png)

after:
![image](https://user-images.githubusercontent.com/27722894/75554650-0c0d7400-5a3b-11ea-9fe9-98dfc0217dca.png)

### Recent plays:

before:
![image](https://user-images.githubusercontent.com/27722894/75554961-a40b5d80-5a3b-11ea-81f5-edb2f0c31d25.png)

after:
![image](https://user-images.githubusercontent.com/27722894/75554692-23e4f800-5a3b-11ea-9fe6-df1f31ad3b85.png)

thewildtree committed:
Make DrawableLinkCompiler use OverlayColourProvider when possible

EVAST9919 commented on osu.Game/Online/Chat/DrawableLinkCompiler.cs:
```diff
@@ -31,10 +32,16 @@ public DrawableLinkCompiler(IEnumerable<Drawable> parts)
+        private OverlayColourProvider colourProvider { get; set; }
+
+        [Resolved]
+        private OsuColour colours { get; set; }
```
Why these have been moved outside of `load()`?

EVAST9919 replied:
Also mapper name colour in Most played beatmaps in UserOverlay is a bit off

TheWildTree replied:
ColourProvider had to be moved out because it can be null (unless you can do that directly in `load()`?) and OsuColour got moved out for consistency, but you're right, it could've stayed where it was

EVAST9919 replied:
you can make stuff nullable in load (check `AppendComments` in `CommentsPage.cs`)

TheWildTree replied:
thanks, I should've checked that earlier. i'll fix that and will look for any other colours being off due to this PR.

bdach replied:
FWIW there's been a PR merged recently that replaced BDL injection with `[Resolved]` wherever possible (#7837). I would like to avoid going back and forth on this stylistic choice.

TheWildTree replied:
yup, they aren't needed anywhere outside the `load()` so i'll move them back there.

thewildtree committed:
Use BDL instead of [Resolved]

thewildtree committed:
Add ability to ignore colour provider in LinkFlowContainer

TheWildTree commented:
Added an option to make `DrawableLinkCompiler` ignore the colour provider because there might be places where we want to keep the links in blue (only occurence I've spotted so far was the `DirectGridPanel`). 

Also adjusted the most played section text - it's now in line with other parts, although still a bit off vs. osu-web because it uses a different text colour for the link there and we can't do that in lazer for now. I think it can stay like that as long as it's consistent with other links.

TheWildTree commented:
Sorry for the force-push, turns out the most played section was a bit off colour-wise so I decided to move the change mentioned above to a separate adjustment PR to avoid merge conflicts here.

bdach commented:
A revert would have been fine, but at least the force doesn't seem to have broken CI (it builds wrong commits sometimes when forcing - I recall it had something to do with it building the latest commit time-wise instead of `BRANCH_HEAD`).

bdach commented:
I kind of dislike the notion of `IgnoreColourProvider` being a property here and my nebulous reasoning is that it feels "too specific". If we want to be able to switch between having theming or no theming in places (which is arguable from a UX/design standpoint, but that is not my domain), I feel like there should be two separate components - `LinkFlowContainer` should mainly stay as-is and a more specific one (`ThemedLinkFlowContainer`? `OverlayLinkFlowContainer`?) would then definitively take in the colour provider and use it without having any extra switching logic inside.

TheWildTree commented:
My original idea was similar, but it was said in the linked issue to keep this simple. It was without taking into account that we'd need to switch between themed and non-themed versions though, so I guess in this situation a specific, themed `LinkFlowContainer` won't be a bad idea. Will see what I can do.

bdach commented:
I mean I would agree with that to the point of a5c004e where that flag wasn't there, but now that we have an injected provider that overrides the colour and then a flag that undoes the override, it seems not too great to me. Let's double check our opinions here first though - @smoogipoo, @peppy, thoughts?

TheWildTree commented:
I thought about this for a while and in my opinion, a separate class for this would indeed be a good idea. Seeing as osu-web uses different link colours in some cases (mostly to make the overall design look more uniform), we are also in need of something that would let us match that lazer-side.

I agree with your initial proposal - `LinkFlowContainer` should IMO stay as-is, while a new one (`ColouredLinkFlowContainer`?) should handle all the non-blue link cases: it would allow us to supply a specific colour to use for the link, or in case of no colour provided (which would mostly be the case) fall back to using `OverlayColourProvider`s `Light2`.

smoogipoo commented:
I agree with the notion of separating it out into a separate class. Either `ColouredLinkFlowContainer` or `OverlayLinkFlowContainer` would be fine imo.

TheWildTree commented:
One more thing - should the yellow colour on hover be kept in all cases or should we allow to customize that as well?

bdach commented:
I'd say the overlay-themed links need to follow theme at all times, and that includes hover colour.

thewildtree committed:
Implement ColouredLinkFlowContainer

TheWildTree commented:
The version I just pushed is probably *far* from perfect - but I couldn't come up with anything better without external help, so I decided to push it here and slowly improve it with the help from more experienced people. 

It does what it's supposed to, though - once you replace a `LinkFlowContainer` with its `Coloured` version, it uses the `OverlayColourProvider` by default while also letting you supply a custom idle / hover colour - I tried my best to make the colour switching work at any moment, both before you add links and after that.

It's not hooked up anywhere yet - I'll do that once it's decided whether this is a good way to implement the desired behaviour.

TheWildTree commented on osu.Game/Online/Chat/ColouredDrawableLinkCompiler.cs:
```diff
@@ -0,0 +1,46 @@
+        {
+            get => hoverColour;
+            set => base.HoverColour = hoverColour = value;
+        }
```
Not too sure about the whole colour setting logic - I didn't want to make these fields from `OsuHoverContainer` public, so I decided to overwrite them this way. Is there a better way to achieve what we need here?

bdach commented:
> It does what it's supposed to, though - once you replace a `LinkFlowContainer` with its `Coloured` version, it uses the `OverlayColourProvider` by default while also letting you supply a custom idle / hover colour - I tried my best to make the colour switching work at any moment, both before you add links and after that.

I'll come back and do a proper review later, but this is entirely unnecessary. Customisation of themed links with public properties should not be allowed, period, unless there is a case of links having different hover/idle colours in different places.

TheWildTree commented:
It kinda is, though - web has (or as least the old design had) a couple places where links were coloured a bit differently. Should we just ignore that then?

bdach commented:
That is extremely annoying. I've already struggled with this a bit in the other PR where ideally I'd have no setters at all and just specify the shades (members of `OverlayColourProvider`) to use declaratively. What I'd like to see in that case is something like this:
```csharp
new ColouredLinkFlowContainer
{
    IdleColour = colourProvider => colourProvider.Light2,
    HoverColour = colourProvider => colourProvider.Light4
}
```
I kind of dislike those being of type `Func<OverlayColourProvider, Color4>` though. Ideally you'd just have an enum. But that's a separate PR for that to be possible.

TheWildTree commented:
That would be great because it'd allow custom colouring while still strictly adhering to the `OverlayColourProvider` scheme, but just as you said, that's out of scope for this one.

I mentioned the non-OCP colouring in one of my earlier comments and implemented that because no one explicitly told me not to. It felt a bit annoying for me too, but oh well - this kind of customization might turn out useful in the future if it goes through.

One thing I'm thinking about is maybe combining the `IdleColour` and `HoverColour` properties into one - it would use the provided colour in the idle state and apply something like `.Lighten()` when hovering to ensure consistent colour differences.

peppy commented:
I think the code in its current state looks fine for now. Not against the proposed direction (Func with colour provider) as a future excursion.

peppy committed:
Merge branch 'master' into coloured-links

peppy committed:
Remove unused usings

peppy commented:
@TheWildTree please go ahead and hook things up again 👍

thewildtree committed:
Don't permit nulls in BDL

thewildtree committed:
Improve colour changing logic

thewildtree committed:
Use ColouredLinkFlowContainer where needed

TheWildTree commented:
Current version turned out to be not-so-fine, I ran into the same colour assignment logic problem as we did a while ago in the `OsuTabControl` (#8117), resulting in the idle / hover colours not being properly assigned if links are added after the colour assignment happens (which is quite often).

I did kind of solve that (by passing current colours to the `ColouredDrawableLinkCompiler` when creating it), but another issue popped up: the base `DrawableLinkCompiler` kept setting the `IdleColour` to blue in its BDL, overriding whatever colour we set when constructing our coloured variant of that class.

My solution to that feels hacky but I couldn't come up with anything better.

I also tried to experiment with using `Bindable`s here, which might've been a cleaner solution to the first problem described above, but couldn't get them to work as intended.

Hooked up the `ColouredLinkFlowContainer` wherever it was needed, with rankings table being an exception since it's being dealt with in another PR and I'll add changes there to avoid merge conflicts.

thewildtree committed:
Simplify condition

thewildtree committed:
Fix dependencies missing in tests

thewildtree committed:
Remove unused variable

bdach commented on osu.Game/Graphics/Containers/ColouredLinkFlowContainer.cs:
```diff
@@ -0,0 +1,54 @@
+                foreach (var linkCompiler in InternalChildren.OfType<ColouredDrawableLinkCompiler>())
+                    linkCompiler.HoverColour = value;
+            }
+        }
```
Yeah, definitely not a fan of these.

> I also tried to experiment with using `Bindable`s here, which might've been a cleaner solution to the first problem described above, but couldn't get them to work as intended.

What was the problem?

TheWildTree replied:
Couldn't get colour value propagation to work properly, though that was probably just my lack of experience with bindables, and I didn't want to spend too much time on something that I wasn't sure will work. I'll try to approach that again later.

bdach replied:
After a short look at the situation the protected-ness of `{Idle,Hover}Colour` is the main problem here. If the base classes do defaults in BDL then there's really no good way to reliably override them. Maybe what is desired here is something that makes the default more explicit, i.e.

```csharp
class OsuHoverContainer
{
        [Resolved]
        private OsuColour colours { get; set; };

        private Color4? hoverColour;

        protected Color4 HoverColour
        {
            get => hoverColour ?? colours.Yellow;
            set => hoverColour = value;
        }

}
```

and the same thing for `IdleColour`. Might be better off as a separate PR, though?
